its_on/admin/views/switches.py

```diff
@@ -140,6 +148,45 @@ async def post(self) -> Dict[str, Dict]:
         raise web.HTTPFound(location=location)
 
 
+class SwitchesCopyAdminView(web.View, CreateMixin):
+    validator = SwitchAddFromAnotherItsOnAdminPostRequestSchema()
+    model = switches
+
+    @staticmethod
+    async def _get_switches_data() -> MultiDictProxy:
+        async with ClientSession() as session:
+            async with session.get(settings.SYNC_FROM_ITS_ON_URL) as resp:
+                return await resp.json()
+
+    @login_required
+    async def post(self) -> None:
+        update_existing = bool(self.request.rel_url.query.get('update_existing'))
+        switches_data = await self._get_switches_data()
+        for switch_data in switches_data['result']:
+            await self._create_or_update_switch(switch_data, update_existing)
+
+        location = self.request.app.router['switches_list'].url_for()
+        raise web.HTTPFound(location=location)
+
+    async def _create_or_update_switch(
+        self, switch_data: MultiDictProxy, update_existing: bool = False,
+    ) -> None:
+        try:
+            await self.create_object(self.request, switch_data)
+        except (ValidationError, psycopg2.IntegrityError):
+            if update_existing:
+                async with self.request.app['db'].acquire() as conn:
```

Comment:
This code snippet looks very cumbersome. I would move this to a private method.

Reply:
Fixed in: https://github.com/best-doctor/its_on/pull/15/commits/7bdd33d84fc2d68073653df9c36b9858c20cedec